### ISSUE TITLE
[BUG] RKE2 provisioning with custom NodeDriver fails when rancher has more than one replica

### ISSUE THREAD

**Issue by michalg91** [rancher/rancher#38481]
**Rancher Server Setup**
- Rancher version: 2.6.6
- Installation option (Docker install/Helm Chart): Helm Chart

**Information about the Cluster**
- Kubernetes version: RKE2  v1.23.7
- Cluster Type (Local/Downstream): Local

**User Information**
- What is the role of the user logged in? Admin

**Describe the bug**

When trying to provision rke2 node with custom docker machine driver (eg. [docker-machine-driver-proxmox-ve](https://github.com/lnxbil/docker-machine-driver-proxmox-ve)), provisioning job fails as it can not find the driver in /assets.
I found out that the driver appears in the /assets directory but only in one rancher instance instead of all in deployment.

**To Reproduce**
Provision rke2 cluster with custom node driver on rancher with more than 1 instance running. 
Putting driver manualy on another instances to /usr/share/rancher/ui/assets/ fixes the issue.

**Result**
```Status:
  Cloud Credential Secret Name:  cattle-global-data:cc-fm9vm
  Conditions:
    Message:  failed creating server [fleet-default/am4-rke2-test-pool1-36e4ded7-r72zm] of kind (ProxmoxveMachine) for machine am4-rke2-test-pool1-77bd49cbf5-sg922 in infrastructure provider: CreateError: Downloading driver from https://rancher.xxx/assets/docker-machine-driver-proxmoxve
Doing /etc/rancher/ssl
downloaded file docker-machine-driver-proxmoxve failed sha256 checksum
download of driver from https://rancher.xxx/assets/docker-machine-driver-proxmoxve failed
    Reason:         CreateError
    Status:         False
    Type:           CreateJob
    Message:        deleting server [fleet-default/am4-rke2-test-pool1-36e4ded7-r72zm] of kind (ProxmoxveMachine) for machine am4-rke2-test-pool1-77bd49cbf5-sg922 in infrastructure provider
    Reason:         
    Status:         False
    Type:           Ready
  Driver Hash:      dd28695b5c454703e9cdb51b0e39719001d8927815f91a03bd1326843fffbd18
  Driver URL:       https://rancher.xxx/assets/docker-machine-driver-proxmoxve
  Failure Message:  Downloading driver from https://rancher.xxx/assets/docker-machine-driver-proxmoxve
Doing /etc/rancher/ssl
downloaded file docker-machine-driver-proxmoxve failed sha256 checksum
download of driver from https://rancher.xxx/assets/docker-machine-driver-proxmoxve failed
```

**Expected Result**
Node should start provisioning and the drivers should appear in /ui/assets in all rancher instances. 

**Screenshots**

**Additional context**
I found out that some of drivers are loaded in dockerfile, probably as the workaround for this issue (https://github.com/rancher/rancher/blob/release/v2.6/package/Dockerfile#L72-L83)

SURE-5893

**Comment by YunSangJun** [comment]
I have a same issue.

**Comment by YunSangJun** [comment]
W/A

Download a custom driver from rancher pod postStart lifecycle. 

```
        lifecycle:
          postStart:
            exec:
              command:
              - sh
              - -c
              - |
                echo "Download custom driver..."
                curl https://... -o /usr/share/rancher/ui/assets/docker-machine-driver-custom
```

**Comment by bashofmann** [comment]
Internal tracking number: SURE-5893

**Comment by jesse-gonzalez** [comment]
Running into the same issue with Rancher 2.7.2+ environment when I deploy min 2-3 Rancher Replicas and try to provision a RKE2 Cluster using Nutanix docker machine driver.  Below are steps to easily reproduce

- [ ] If Rancher is already running, scale down replicas to ZERO and then back up to 3
- [ ] Deploy cluster with min. 2 Nodes/NodePools and immediately tail machine logs in `fleet-default` namespace

```bash
## download success
rke2-source-cluster-13690-rke2-source-cluster-13690-serve-rv849 machine Downloading driver from https://rancher.<xyz.com>/assets/docker-machine-driver-nutanix
rke2-source-cluster-13690-rke2-source-cluster-13690-serve-rv849 machine Doing /etc/rancher/ssl
rke2-source-cluster-13690-rke2-source-cluster-13690-serve-rv849 machine docker-machine-driver-nutanix
rke2-source-cluster-13690-rke2-source-cluster-13690-serve-rv849 machine docker-machine-driver-nutanix: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, stripped
rke2-source-cluster-13690-rke2-source-cluster-13690-serve-rv849 machine About to remove rke2-source-cluster-13690-rke2-source-cluster-13690-serve-k4zsp
rke2-source-cluster-13690-rke2-source-cluster-13690-serve-rv849 machine WARNING: This action will delete both local reference and remote instance.
rke2-source-cluster-13690-rke2-source-cluster-13690-serve-rv849 machine Machine config for rke2-source-cluster-13690-rke2-source-cluster-13690-serve-k4zsp does not exists, so nothing to do...

## download failed
rke2-source-cluster-13690-rke2-source-cluster-13690-agent-6qbz7 machine Downloading driver from https://rancher.<xyz.com>/assets/docker-machine-driver-nutanix
rke2-source-cluster-13690-rke2-source-cluster-13690-agent-6qbz7 machine Doing /etc/rancher/ssl
rke2-source-cluster-13690-rke2-source-cluster-13690-agent-6qbz7 machine downloaded file docker-machine-driver-nutanix failed sha256 checksum
rke2-source-cluster-13690-rke2-source-cluster-13690-agent-6qbz7 machine download of driver from https://rancher.<xyz.com>/assets/docker-machine-driver-nutanix failed

```

- [ ] Navigate to https://rancher.<xyz.com>/assets/ and see if `docker-machine-driver-nutanix` exists. Refresh browser multiple times and see that driver is not available after being load balanced between rancher endpoints.

- [ ] Check all available pods for expected file and see that only one pod has available

```bash
$ kubectl get po -l app=rancher -o name -n cattle-system | xargs -t -I {} kubectl exec {} -n cattle-system -- ls -l /usr/share/rancher/ui/assets/docker-machine-driver-nutanix

kubectl exec pod/rancher-795c9776d4-cbphd -n cattle-system -- ls -l /usr/share/rancher/ui/assets/docker-machine-driver-n
utanix
ls: cannot access '/usr/share/rancher/ui/assets/docker-machine-driver-nutanix': No such file or directory
command terminated with exit code 2

kubectl exec pod/rancher-795c9776d4-kcj8m -n cattle-system -- ls -l /usr/share/rancher/ui/assets/docker-machine-driver-n
utanix
ls: cannot access '/usr/share/rancher/ui/assets/docker-machine-driver-nutanix': No such file or directory
command terminated with exit code 2

kubectl exec pod/rancher-795c9776d4-lkz4z -n cattle-system -- ls -l /usr/share/rancher/ui/assets/docker-machine-driver-n
utanix
-rwxr-xr-x 1 root root 8749056 May  9 20:30 /usr/share/rancher/ui/assets/docker-machine-driver-nutanix
```

**Comment by jiaqiluo** [comment]
### Root Cause:

When multiple rancher pods exist, one becomes the leader, and the custom node driver is only downloaded in the leader pod.  This could explain the behavior [in this comment](https://github.com/rancher/rancher/issues/38481#issuecomment-1540893657) that the `/usr/share/rancher/ui/assets/docker-machine-driver-nutanix` binary exists only in one of three rancher pods. 

When `rancher-machine`  tries to download the driver via the endpoint `rancher.<xyz.com>/assets/docker-machine-driver-nutanix`, the request is load balanced to one of the three pods randomly; when it is sent to a non-leader pod in which the binary does not exist, a `404 not found` error is returned, therefore the sha256 checksum fails. ([code](https://github.com/rancher/machine/blob/2c59804f13cd37d2faad2017047c1d40264e0674/package/download_driver.sh#L20-L26))

**Comment by markusewalker** [comment]
**QA TEST PLAN**

**Scenarios**
|Scenario|Test Case|
|---|---|
|1|Setup Rancher with 1 replica using Proxmox VE, provision a downstream RKE2 node driver cluster as a standard user, validate that the docker machine driver appears in the /assets directory|
|2|Setup Rancher with 3 replicas using Proxmox VE, provision a downstream RKE2 node driver cluster as a standard user, validate that the docker machine driver appears in the /assets directory|

**Comment by markusewalker** [comment]
Tested this on `v2.7.5` and `v2.7-head` with the same unsuccessful result. See details below:

**ENVIRONMENT DETAILS**
- Rancher install: Helm
- Rancher version: `v2.7.5` and `v2.7-head` https://github.com/rancher/rancher/commit/369c7d37b9e678922297c3811b4329e18da4abdf
- Browser: Chrome
- Proxmox VE

**QA TEST PLAN**

**Scenarios**
|Scenario|Test Case|
|---|---|
|1|Setup Rancher with 3 replica using Proxmox VE, provision a downstream RKE2 node driver cluster as a standard user, validate that the docker machine driver appears in the /assets directory|

**TEST RESULT**
|Scenario|Test Case|Result
|---|---|---|
|1|Setup Rancher with 3 replica using Proxmox VE, provision a downstream RKE2 node driver cluster as a standard user, validate that the docker machine driver appears in the /assets directory|:x:|

**VALIDATION STEPS**

**Scenario 1**
1. Setup Proxmox and created an K3S cluster with one main node and one agent node.
2. Installed Rancher with 3 replicas and accessed it in a UI with a standard user.
3. Provisioned a downstream RKE2 `v1.26.6+rke2r1` AWS node driver cluster.
4. Noted that I am STUCK in the following step:
![image](https://github.com/rancher/rancher/assets/52591819/b1b80f18-4c52-4a83-ae85-8ac0ace400d3) 
     - Troubleshooting further, I noted that the created AWS nodes are unable to speak with the Proxmox VM. So this is most likely an unrelated networking error and not a provisioning issue.
5. To validate the above, attempted to provision a custom RKE2 `v1.26.6+rke2r1` cluster using the main Proxmox VM.
     - I was not stuck in the `Reconciling` state and the cluster was waiting for kubelet and the other tools.
     - I did not see provisioning succeed, but I had this as more of a proof of concept for the above networking issue.

**NOTES/COMMENTS**
While the aforementioned networking issue does not appear to be an issue for further testing, there have been multiple complications with this issue in testing. As such, reopening this issue to speak internally on next steps for QA to confidently test this issue.

**Comment by michalg91** [comment]
Hi. I think QA test is not testing the issue. The problem is about
provisioning using custom node driver not aws driver. So if you have
proxmox ve installed add
https://github.com/lnxbil/docker-machine-driver-proxmox-ve to nodedrivers
in rancher ui and create cluster using it.

czw., 27 lip 2023, 00:28 użytkownik Markus Walker ***@***.***>
napisał:

**Comment by markusewalker** [comment]
**UPDATED QA TEST PLAN**

**Scenarios**
|Scenario|Test Case|
|---|---|
|1|Setup Rancher with 3 replicas, add custom driver Proxmox VE, provision a downstream RKE2 custom node driver cluster as a standard user, validate that the docker machine driver appears in the /assets directory|
|2|Setup Rancher v2.7.5 with 3 replicas, add custom driver Proxmox VE, provision a downstream RKE2 custom node driver cluster as a standard user, upgrade to Rancher v2.7-head, provision a new downstream RKE2 custom node driver cluster as a standard user, validate that the docker machine driver appears in the /assets directory|

**Comment by markusewalker** [comment]
Reproduced this on `v2.7.5`. See details below:

**ENVIRONMENT DETAILS**
- Rancher install: Helm
- Rancher version: `v2.7.5`
- Browser: Chrome
- Proxmox VE

**TEST RESULT**
|Scenario|Test Case|Result
|---|---|---|
|1|Setup Rancher with 3 replicas, add custom driver Proxmox VE, provision a downstream RKE2 custom node driver cluster as a standard user, validate that the docker machine driver appears in the /assets directory|:x:|

**REPRO STEPS**

**Scenario 1**
1. Setup a separate Proxmox server and a separate Rancher server with 3 replicas.
2. As an admin user, added custom node [driver](https://github.com/lnxbil/docker-machine-driver-proxmox-ve) for Proxmox VE.
     - Due to this [issue](https://github.com/rancher/rancher/issues/37074), I had to run the following command in my client machine: `kubectl get nodedriver <listed Proxmox name> -o yaml > driver.yaml`
     - Excluded everything underneath `metadata` except name and inputted `proxmoxve`. Additionally, removed the `status` block.
     - Ran command `kubectl apply -f driver.yaml`.
3. Noted that the correct Proxmox VE shows in the node driver list in the Rancher UI.
4. Added the following annotations to the custom node driver:
```
privateCredentialFields: proxmoxUserPassword
publicCredentialFields: proxmoxUserName
```
5. As a standard user, attempted to provision a downstream RKE2 `v1.26.6+rke2r1` custom node driver cluster and was unable to.
6. Noted as well that  the driver DID NOT appear in the `/assets` directory as originally reported.

Now that this is reported, QA will attempt to test on the latest `v2.7-head` and see if this issue has been addressed.

**Comment by nnoethling** [comment]
Please have a look into this as this does affect multiple rancher customers

**Comment by markusewalker** [comment]
Reproduced this on `v2.7.5`. See details below:

**ENVIRONMENT DETAILS**
- Rancher install: Helm
- Rancher version: `v2.7-head` https://github.com/rancher/rancher/commit/e0e54ea8672eec2617a58ca701a9224e49d38772
- Browser: Chrome

**TEST RESULT**
|Scenario|Test Case|Result
|---|---|---|
|1|Setup Rancher with 3 replicas, add custom driver Nutanix, provision a downstream RKE2 custom node driver cluster as a standard user|:x:

**VALIDATION STEPS**

**Scenario 1**
1. Setup Rancher via Helm and ensured 3 replicas are available.
2. As a standard user, attempted to provision a downstream RKE2 Nutanix node driver `v1.26.7+rke2r1`  cluster (1 etcd, 1 cp, 1 worker).
3. Received the following error message:
![image](https://github.com/rancher/rancher/assets/52591819/06346d91-b7d7-4843-8ee3-f36dfe4cc782)

Now that this is reproduced, I will be testing on the latest `v2.7-head` to validate that this is addressed.

**Comment by markusewalker** [comment]
Validated that this is addressed on `v2.7-head` https://github.com/rancher/rancher/commit/e0e54ea8672eec2617a58ca701a9224e49d38772. See details below:

**ENVIRONMENT DETAILS**
- Rancher install: Helm
- Rancher version: `v2.7-head` https://github.com/rancher/rancher/commit/e0e54ea8672eec2617a58ca701a9224e49d38772
- Browser: Chrome

**TEST RESULT**
|Scenario|Test Case|Result
|---|---|---|
|1|Setup Rancher with 3 replicas, add custom driver Nutanix, provision a downstream RKE2 custom node driver cluster as a standard user|:white_check_mark:|
|2|Setup Rancher with 3 replicas, add custom driver Nutanix, provision a downstream RKE2 custom node driver cluster as a standard user, upgrade the K8s version|:white_check_mark:|
|3|Setup Rancher v2.7.5 with 3 replicas, add custom driver Nutanix, provision a downstream RKE2 custom node driver cluster as a standard user, upgrade to Rancher v2.7-head, provision a new downstream RKE2 custom node driver cluster as a standard user|:white_check_mark:|

**VALIDATION STEPS**

**Scenario 1**
1. Setup Rancher via Helm and ensured 3 replicas are available.
2. As a standard user, provisioned a downstream RKE2 Nutanix node driver `v1.26.7+rke2r1`  cluster (1 etcd, 1 cp, 1 worker).
3. Noted that the cluster is stuck at `Waiting for cluster agent to connect`, but the reported issue has been addressed.
     - Root causing the issue has shown this to be a DNS issue with Nutanix as opposed to a Rancher issue.
     - As such, marking this test case as passing as this unrelated DNS issue is causing the cluster to not be `Active`.
4. Repeated this scenario using a Docker install instead of Helm; the cluster and nodes DID come up as `Active`.
     - The only difference was the Rancher server URL used the corresponding IP address and not it's specified DNS. This confirms that the DNS is the issue.
________

**Scenario 2**
1. Repeated scenario 1, but provisioned a downstream RKE2 Nutanix node driver `v1.25.17+rke2r1` cluster (1 etcd + cp, worker).
2. Received the same results as scenario 1.
3. Upgraded the downstream cluster to `v1.25.6+rke2r1`.
4. Ran into the above-mentioned DNS issue, so the cluster remained in `Waiting for cluster agent to connect` status.
5. Repeated this scenario using a Docker install instead of Helm; the cluster and nodes DID come up as `Active`.
     - The only difference was the Rancher server URL used the corresponding IP address and not it's specified DNS. This confirms that the DNS is the issue.

__________

**Scenario 3**
1. Setup Rancher `v2.7.5` via Helm and ensured 3 replicas are available.
2. As a standard user, provisioned a downstream RKE2 Nutanix node driver `v1.26.7+rke2r1`  cluster (1 etcd, 1 cp, 1 worker).
3. Noted that the cluster is stuck at `Waiting for cluster agent to connect`, but the reported issue has been addressed.
4. Upgraded Rancher to `v2.7-head`; validated that the cluster did not auto-upgrade the K8s version.
5. Provisioned a new downstream RKE2 Nutanix node driver `v1.26.7+rke2r1`  cluster (1 etcd, 1 cp, 1 worker).
     - Noted the same DNS issue as seen, leaving the cluster in state `Waiting for cluster agent to connect`.
